lib/ansible/utils/_ssh_agent.py

```diff
@@ -0,0 +1,671 @@
+                return Ed25519PublicKeyMsg
+            case KeyAlgo.DSA:
+                return DSAPublicKeyMsg
+            case _:
+                raise NotImplementedError(type)
+
+    def public_key(self) -> CryptoPublicKey:
+        type: KeyAlgo = self.type
+        match type:
+            case KeyAlgo.RSA:
+                return RSAPublicNumbers(
+                    self.e,
+                    self.n
+                ).public_key()
+            case KeyAlgo.ECDSA256 | KeyAlgo.ECDSA384 | KeyAlgo.ECDSA521:
+                curve = _ECDSA_KEY_TYPE[KeyAlgo(type)]
+                return EllipticCurvePublicKey.from_encoded_point(
+                    curve(),
+                    self.Q
+                )
+            case KeyAlgo.ED25519:
+                return Ed25519PublicKey.from_public_bytes(
+                    self.enc_a
+                )
+            case KeyAlgo.DSA:
+                return DSAPublicNumbers(
+                    self.y,
+                    DSAParameterNumbers(
+                        self.p,
+                        self.q,
+                        self.g
+                    )
+                ).public_key()
+            case _:
+                raise NotImplementedError(type)
+
+    @staticmethod
+    def from_public_key(public_key: CryptoPublicKey) -> PublicKeyMsg:
+        match public_key:
+            case DSAPublicKey():
+                dsa_pn: DSAPublicNumbers = public_key.public_numbers()
+                return DSAPublicKeyMsg(
+                    KeyAlgo.DSA,
+                    mpint(dsa_pn.parameter_numbers.p),
+                    mpint(dsa_pn.parameter_numbers.q),
+                    mpint(dsa_pn.parameter_numbers.g),
+                    mpint(dsa_pn.y)
+                )
+            case EllipticCurvePublicKey():
+                return EcdsaPublicKeyMsg(
+                    getattr(KeyAlgo, f'ECDSA{public_key.curve.key_size}'),
+                    unicode_string(f'nistp{public_key.curve.key_size}'),
+                    binary_string(public_key.public_bytes(
+                        encoding=serialization.Encoding.X962,
+                        format=serialization.PublicFormat.UncompressedPoint
+                    ))
+                )
+            case Ed25519PublicKey():
+                return Ed25519PublicKeyMsg(
+                    KeyAlgo.ED25519,
+                    binary_string(public_key.public_bytes(
+                        encoding=serialization.Encoding.Raw,
+                        format=serialization.PublicFormat.Raw,
+                    ))
+                )
+            case RSAPublicKey():
+                rsa_pn: RSAPublicNumbers = public_key.public_numbers()
+                return RSAPublicKeyMsg(
+                    KeyAlgo.RSA,
+                    mpint(rsa_pn.e),
+                    mpint(rsa_pn.n)
+                )
+            case _:
+                raise NotImplementedError(public_key)
+
+    def fingerprint(self) -> str:
```

Comment:
This should probably be a cached property too.

Reply:
Done.